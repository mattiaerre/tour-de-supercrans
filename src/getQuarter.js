import moment from 'moment';

function getQuarter(input) {
  const now = moment(input);
  const month = now.month();
  const year = now.year();

  let first;
  let last;
  let text;
  switch (month) {
    case 0:
    case 1:
    case 2:
      first = '01';
      last = '03';
      text = 'Q1';
      break;
    case 3:
    case 4:
    case 5:
      first = '04';
      last = '06';
      text = 'Q2';
      break;
    case 6:
    case 7:
    case 8:
      first = '07';
      last = '09';
      text = 'Q3';
      break;
    default:
      first = '10';
      last = '12';
      text = 'Q4';
      break;
  }

  const quarter = {
    first: moment(`${year}-${first}-01`).format('YYYY-MM-DD'),
    last: moment(`${year}-${last}-15`).endOf('month').format('YYYY-MM-DD'),
    text,
    year
  };

  return quarter;
}

export default getQuarter;
