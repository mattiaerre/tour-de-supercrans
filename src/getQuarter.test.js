import getQuarter from './getQuarter';

describe('getQuarter', () => {
  const scenarios = [
    { day: '2020-01-14', description: 'Q1', expected: 'Q1' },
    { day: '2020-05-05', description: 'Q2', expected: 'Q2' },
    { day: '2020-09-29', description: 'Q3', expected: 'Q3' },
    { day: '2020-12-25', description: 'Q4', expected: 'Q4' },
    { day: '2023-04-15', description: 'Q2', expected: 'Q2' },
    { day: '2023-07-15', description: 'Q3', expected: 'Q3' },
    { day: '2023-08-15', description: 'Q3', expected: 'Q3' }
  ];

  scenarios.forEach(({ day, description, expected }) => {
    test(description, () => {
      const quarter = getQuarter(day);
      expect(quarter.text).toEqual(expected);
      expect(quarter).toMatchSnapshot();
    });
  });

  describe('Fiscal year', () => {
    const scenarios = [
      {
        day: '2022-10-26',
        description: 'FY2023/Q2',
        expected: { text: 'Q2', year: 2023 }
      },
      {
        day: '2023-02-02',
        description: 'FY2023/Q3',
        expected: { text: 'Q3', year: 2023 }
      }
    ];

    scenarios.forEach(({ day, description, expected }) => {
      test(description, () => {
        const quarter = getQuarter(day);
        expect(quarter.fiscalText).toEqual(expected.text);
        expect(quarter.fiscalYear).toEqual(expected.year);
        expect(quarter).toMatchSnapshot();
      });
    });
  });
});
