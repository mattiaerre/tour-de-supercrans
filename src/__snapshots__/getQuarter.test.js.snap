// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQuarter Fiscal year FY2023/Q2 1`] = `
Object {
  "first": "2022-10-01",
  "fiscalText": "Q2",
  "fiscalYear": 2023,
  "last": "2022-12-31",
  "text": "Q4",
  "year": 2022,
}
`;

exports[`getQuarter Fiscal year FY2023/Q3 1`] = `
Object {
  "first": "2023-01-01",
  "fiscalText": "Q3",
  "fiscalYear": 2023,
  "last": "2023-03-31",
  "text": "Q1",
  "year": 2023,
}
`;

exports[`getQuarter Q1 1`] = `
Object {
  "first": "2020-01-01",
  "fiscalText": "Q3",
  "fiscalYear": 2020,
  "last": "2020-03-31",
  "text": "Q1",
  "year": 2020,
}
`;

exports[`getQuarter Q2 1`] = `
Object {
  "first": "2020-04-01",
  "fiscalText": "Q4",
  "fiscalYear": 2020,
  "last": "2020-06-30",
  "text": "Q2",
  "year": 2020,
}
`;

exports[`getQuarter Q3 1`] = `
Object {
  "first": "2020-07-01",
  "fiscalText": "Q1",
  "fiscalYear": 2021,
  "last": "2020-09-30",
  "text": "Q3",
  "year": 2020,
}
`;

exports[`getQuarter Q4 1`] = `
Object {
  "first": "2020-10-01",
  "fiscalText": "Q2",
  "fiscalYear": 2021,
  "last": "2020-12-31",
  "text": "Q4",
  "year": 2020,
}
`;
